### ISSUE TITLE
Back propagation derivatives 

### ISSUE THREAD

**Issue by flare9x** [makeyourownneuralnetwork/makeyourownneuralnetwork#64]
This implementation of back propagation in the book is actually a lot different to some other methods I have come across. It does seem specific to the sigmoid function?

In the book: 
```

# output layer error is the (target - actual)
    output_errors = targets - final_outputs
    # hidden layer error is the output_errors, split by weights, recombined at hidden nodes
    hidden_errors = whoT * output_errors

    # update the weights for the links between the hidden and output layers
    who .= who .+ (lr .* (output_errors .* final_outputs .* (1.0 .- final_outputs)) * hidden_outputsT)

    # update the weights for the links between the input and hidden layers
    wih .= wih .+ (lr .* (hidden_errors .* hidden_outputs .* (1.0 .- hidden_outputs)) * inputsT)

# Note .* is element wise multiplication and .- is element wise subtraction ( * ) is matrix multiplication 
```
In the post below:

https://zhenye-na.github.io/2018/09/09/build-neural-network-with-mnist-from-scratch.html#loss-function

The backpropogation is as below: This also accounts for different activation functions so long as you remember to apply the derivative of the activation function during back propogation:
```

def back_propagate(X, Y, params, cache, m_batch):
    """
    back propagation

    inputs:
        params: dictionay a dictionary contains all the weights and biases
        cache: dictionay a dictionary contains all the fully connected units and activations

    return:
        grads: dictionary a dictionary contains the gradients of corresponding weights and biases
    """
    # error at last layer
    dZ2 = cache["A2"] - Y

    # gradients at last layer (Py2 need 1. to transform to float)
    dW2 = (1. / m_batch) * np.matmul(dZ2, cache["A1"].T)
    db2 = (1. / m_batch) * np.sum(dZ2, axis=1, keepdims=True)

    # back propgate through first layer
    dA1 = np.matmul(params["W2"].T, dZ2)
    dZ1 = dA1 * sigmoid(cache["Z1"]) * (1 - sigmoid(cache["Z1"])) 

#-------------------------
# apply derivative of your activation function here!! ie the above line then becomes:
dZ1 = dA1 * ReLU_derivative(cache["Z1"])) OR 
dZ1 = dA1 * leaky_ReLU_derivative(cache["Z1"]))
#---------------------------

    # gradients at first layer (Py2 need 1. to transform to float)
    dW1 = (1. / m_batch) * np.matmul(dZ1, X.T)
    db1 = (1. / m_batch) * np.sum(dZ1, axis=1, keepdims=True)

    grads = {"dW1": dW1, "db1": db1, "dW2": dW2, "db2": db2}

    return grads
```

I wish I 100% understood the difference between the two!

On further inspection your implementation on MNIST achieves a 98% training accuracy over 5 epochs. Test data 97%.
The second method I posted in 5 epochs achieves 92% accuracy. And during training seems to saturate or stagnate around 88% area. Although it does a quick enough job getting there. I feel something is off with the second implementation. 

Really do not understand why your one is more superior! 

Sorry if I didnt leave high school right after GCSE i may have a better handle on derivatives. Mechanically learning through the code.

**Comment by makeyourownneuralnetwork** [comment]
hi - good question!

There are many options for loss functions and activation functions.

I stick to very simple ones because my aim is to maximise understanding for readers, with minimal distractions.

The cross entropy loss is useful when you know there should only be "one output" - but it also has it's downsides in that it can make the network overly confident. 

The ReLU activation is good because it avoids what can happen with the normal sigmoid where the network saturates with high values which then lead to very small gradients to learn from.
You should get good results with this - so I'm not sure why you're not! Are you using a framework like tensorflow or pytorch?

dont' worry about leaving school - you clearly have an interest and you understand the key ideas - keep going!

**Comment by flare9x** [comment]
Thanks Tariq.
I actually started to build the neural network with julia language (fast!)
and the cnn with julia but thinking today that perhaps its not all that
important that i can build them from scratch rather than refine the
implementation of them on real world problems and data. So with that being
said, looking to begin using tensorflow and brushing up on python!

Many thanks
Andrew
